Adapted Classifiers

walragatver opened the pull request:
Hi, 
This is pull Request for #1483 Issue.
I have changed three classifiers 
1) adaboost - returning ztProduct as it's upper bound for training error
2) decision stump - returning the entropy after splitting.
3) decision tree - returning final entropy of the tree . Added some code to get final entropy in recursive decision   tree.
Let me know if I have returned something wrong.

Update One:
Changed Two Classifiers 
1) HMM - In this returning final log likelihood obtained after training using data sequences.
2) Random Forest  - Returning average entropy of all the decision trees trained under the forest
                                 (Not Sure of this)

Update Two:
1) LARS : Returning final absolute maximum correlation after training
2) Linear_regression : Returning least squares error after training.
3) sparse_coding : Returning final minmized objective function value.
4) ANN : Returning objective function from GANS, FNN, RNN, RBM.  

Update Three:
1) local_coordinate_coding : returning final objective value

Also I went through some other classifiers and drawn following conclusions
1) Hoeffding Trees : In this we are using gain to find best split hence returning entropy is not good.
2) Naive Bayes Classifier : In this we are updating the probability and distribution curve and thus we don't need to return anything in this.
3) Perceptron : In this we are doing simple weight update without any objective function hence no need to return anything from it.
4) Softmax Regression : Train function is already modified.
5) GMM : Train function is already modified


I am quite new to these algorithms, please correct me if I have made a mistake .

walragatver commented:
Hi @rcurtin, @zoq,
    Sorry for disturbance. But it looks like I have covered all classifiers. Are the conclusions drawn by me correct?

rcurtin commented:
Hey @walragatver, sorry that I haven't had a chance to take a look yet.  I'll review it when I'm able to (hopefully in the next few days). :+1:

zoq commented:
Sorry for the slow response, will take a look at the PR over the weekend.

rcurtin commented on src/mlpack/methods/decision_tree/decision_tree_impl.hpp:
```diff
@@ -656,9 +656,24 @@ void DecisionTree<FitnessFunction,
+          double newContribute = newChildGain * childCounts[i] / count;
+          newBestGain = newBestGain + newContribute;
+          bestGain = std::max(bestGain,newBestGain);
+        }
```
It looks like the strategy you're using here is to check if the entropy of the child is different than the entropy of the child set.  That might happen if the child node makes any additional splits.  But when we call `child->Train()`, it will return either exactly the value of `prevChildGain`, if no further splits were done (i.e. if that child was a leaf), or something better, if further splits were done.  So there should be no need to compute or store `prevChildGain`.

Also, it looks like your strategy for computing `bestGain` is to subtract the contribution of the child points from the original `bestGain` (computed at the beginning of the method), and then add back in the contribution of the newly-built child node.  I think we can ignore the original `bestGain` computation though and essentially do something like this:

```
// Outside of the loop, initialize bestGain.
bestGain = 0;

// This is the loop over the children.
for (size_t i = 0; i < numChildren; ++i)
{
  ...
  double childGain = child->Train(...);
  bestGain += double(childCounts[i]) / double(count) * childGain;
}
```

(The `double()` is to make sure we don't do integer division.  I think it's necessary but I'm not 100% sure.)

Let me know if I can clarify this idea further.

walragatver replied:
Yes, I got that thanks for the suggestion

rcurtin commented on src/mlpack/methods/lars/lars.cpp:
```diff
@@ -129,7 +129,7 @@ void LARS::Train(const arma::mat& matX,
     lambdaPath[0] = lambda1;
     Timer::Stop("lars_regression");
-    return;
+    return maxCorr;
```
I'm not sure that `maxCorr` is the objective minimized by LARS.  I think that objective is Equation 1.3 from https://projecteuclid.org/download/pdfview_1/euclid.aos/1083178935 .  Maybe I overlooked how `maxCorr` is equivalent to that value?

walragatver replied:
Actually, I was not sure of that. 
I am returning that because in section 2 fig 3(right) of the paper they have plotted absolute current correlations as function of LARS step . In that maximum correlation decreases as we keep training. So I thought it's quite similar to decreasing error hence I thought returning that might be reasonable. Also I went through code and it was training on maxCorr so I thought that returning maxCorr is correct. Also in our code there is no function or something which is calculating that error. 
If there are modifications let me know

rcurtin replied:
I think there is a relation between the two---we keep adding variables to the solution until `maxCorr` is sufficently small, so in some sense `maxCorr` captures the amount of `Y` that is not captured by `X*beta`.  There may be some way to convert to something much closer to the mean-squared-error of the regression, but if not I think `maxCorr` can work.  If you'd like to take the time to look into it further, please feel free.

rcurtin commented on src/mlpack/methods/linear_regression/linear_regression.cpp:
```diff
@@ -85,6 +85,7 @@ void LinearRegression::Train(const arma::mat& predictors,
       lambda * arma::eye<arma::mat>(p.n_rows, p.n_rows);
 
   parameters = arma::solve(cov, p * r.t());
+  return ComputeError(predictors,responses);
```
Tiny style issue---there should be a space between `predictors,` and `responses`.

walragatver replied:
Okay fixed it!

rcurtin commented on src/mlpack/methods/random_forest/random_forest_impl.hpp:
```diff
@@ -431,6 +431,7 @@ void RandomForest<
 {
   // Train each tree individually.
   trees.resize(numTrees); // This will fill the vector with untrained trees.
+  double avgGain = 0.0;
```
I think we might want to use an OpenMP reduce or something like this.  As it's written, `avgGain` is shared across all threads in the parallel loop for training each tree, so there may be locking issues.

walragatver replied:
Done !

rcurtin replied:
:+1:

rcurtin commented on src/mlpack/methods/sparse_coding/sparse_coding_impl.hpp:
```diff
@@ -106,6 +106,7 @@ void SparseCoding::Train(
   }
 
   Timer::Stop("sparse_coding");
+  return lastObjVal;
```
Should this be `curObjVal`?

walragatver replied:
No actually, curObjVal is only in the scope of for loop , so we cannot return that. But lastObjVal is in the function scope and in the for loop we can see that we are copying curObjVal in lastObjVal (line 105). So I thought instead of declaring curObjVal outside for loop. It's better to return lastObjVal as it's getting updated with curObjVal at the end of for loop.

rcurtin replied:
Ok, but in this case take a look at line 101---if it converges, then `lastObjVal` is not up to date.  So I guess we need to update it there.

walragatver commented on src/mlpack/methods/adaboost/adaboost.hpp:
```diff
@@ -141,7 +141,7 @@ class AdaBoost
    * @param learner Learner to use for training.
    */
-  void Train(const MatType& data,
+  double Train(const MatType& data,
```
Okay Fixed it !

walragatver commented on src/mlpack/methods/adaboost/adaboost_impl.hpp:
```diff
@@ -198,6 +198,7 @@ void AdaBoost<WeakLearnerType, MatType>::Train(
     // Accumulate the value of zt for the Hamming loss bound.
     ztProduct *= zt;
   }
+  return ztProduct;
```
Yes, It looks like many of our tests can be modified after this modification. I am thinking to work on this in a separate PR

rcurtin replied:
Ok, that's fine, but could you please open an issue so that we don't forget it?  You can assign it to yourself if you like.

walragatver commented:
Hi @rcurtin ,
I have fixed and answered the things you have said. Just let me know if I have made a mistake.
Also I have modified local_coordinate_coding.
Umm, actually I have not adapted searching algorithms like range_search and collabarative filtering. I will adapt them in 2 days and will ping you for final review.

rcurtin commented:
Range search doesn't have a reasonable objective so I am not sure what we would return there.  If you can think of something reasonable, go ahead, but I'm not sure what.  (Similar situation for collaborative filtering.)

walragatver commented:
Hi @rcurtin ,
I have checked all the above algorithms and it looks like we cannot return anything reasonable. It looks like most of them use matrices as it's output. Also your thesis on dual tree algorithms helped me a lot in clearing the basics of dual and single-tree algorithms. 
So, this PR is now ready for it's final review.

rcurtin commented on src/mlpack/methods/decision_tree/decision_tree_impl.hpp:
```diff
@@ -624,7 +624,7 @@ void DecisionTree<FitnessFunction,
 
     // Figure out counts of children.
-    arma::Row<size_t> childCounts(numClasses, arma::fill::zeros);
+    arma::Row<size_t> childCounts(numChildren, arma::fill::zeros);
```
Yeah.  I think we happen to be guaranteed that `numClasses` is always greater than or equal to `numChildren` in the code we have, so it was a dormant bug, but still a bug.  Nice catch :sunglasses:

rcurtin commented on src/mlpack/methods/decision_tree/decision_tree_impl.hpp:
```diff
@@ -624,10 +626,16 @@ void DecisionTree<FitnessFunction,
       childCounts[childAssignments[i - begin]]++;
 
+    //Initialize bestGain if recursive split is allowed
+    if(!NoRecursion)
```
Just a few style comments---can you make sure that each comment is a fully punctuated sentence?  (i.e. period at the end).  Also we can put a space between `//` and `Initialize`, and a space between `if` and `(!NoRecursion)` to match the style guide.  These comments apply throughout, by the way, even though I've only pointed them out here.

rcurtin commented on src/mlpack/methods/decision_tree/decision_tree_impl.hpp:
```diff
@@ -675,6 +685,7 @@ void DecisionTree<FitnessFunction,
         numClasses,
         UseWeights ? weights.subvec(begin, begin + count - 1) : weights);
   }
+  return -bestGain;
```
Do we need to negate a second time here?

walragatver replied:
Yes, if as we can see we are maximizing entropy. Objective function always minimizes something. So returning negative entropy of final tree is reasonable.

rcurtin replied:
Sorry, I thought that it was avoidable since we're already negating `childGain` earlier, but I realize that the code you have here also handles the case where no split is made (in which case `bestGain` needs to be negated).

walragatver commented:
Hi @rcurtin, @zoq,
I have made the final changes and fixed some static errors. It looks there is some glitch in the static check because the error it is pointing is not an error. There is no local `activeSet` in the train function of `lars.cpp`. 
Can you look into it?

rcurtin commented:
Can you try rebasing the master branch into your branch?  #1713 had some fixes for the static code analysis checks that might help here.

walragatver commented:
@rcurtin: I have already rebased my branch but it looks like a new error. I tried to clear it but looks like it's not an error.

rcurtin commented:
Ah, ok, sorry.  Yeah, I took a look at the code and I have to agree, the static code analyzer must be wrong here.  I can't see why it would think that `activeSet` is local.  So let's just not worry about that. :+1:

walragatver commented:
Hi @rcurtin, @zoq,
Can you review this PR today? I am thinking to move towards writing and test cases and finish it as fast as possible. 
Also I went through this year's GSOC ideas they are quite fascinating. Which one will you recommend for me?

rcurtin commented:
It's in my queue to do soon, but this week is a paper deadline week so there's not too much time for me.  Sorry about that...

zoq commented:
@walragatver regarding GSoC, difficult to recommend something at this point, personally, I would go with the idea I'm most passionate about. Also, you are free to propose something that's not listed.

zoq commented on src/mlpack/methods/ann/ffn_impl.hpp:
```diff
@@ -84,12 +84,13 @@ void FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Train(
 
   Log::Info << "FFN::FFN(): final objective of trained model is " << out
       << "." << std::endl;
+  return out;
```
Might be more intuitive to call this loss, or objective, what do you think?

walragatver replied:
Yes, correct but it was already declared or defined. So I thought it's better not to change it's name. Also at every optimizer function it was written like that so I decided to keep it as it is.

zoq replied:
I see, I don't mind to leave it like this.

zoq commented on src/mlpack/methods/ann/ffn.hpp:
```diff
@@ -97,11 +97,12 @@ class FFN
    * @param predictors Input training variables.
    * @param responses Outputs results from input training variables.
    * @param optimizer Instantiated optimizer used to train the model.
+   * @return final objective value.
```
Picky style comment, this should start with a capital letter.

walragatver replied:
Done

walragatver commented:
Hi @zoq, 
I have made the required changes.

zoq commented on src/mlpack/methods/ann/rbm/rbm_impl.hpp:
```diff
@@ -50,6 +50,7 @@ RBM<InitializationRuleType, DataType, PolicyType>::RBM(
     reset(false)
 {
   numFunctions = this->predictors.n_cols;
+  steps = 0;
```
Do you mind to use the member initializer list for the ``steps`` parameter for consistency?

rcurtin commented on src/mlpack/methods/linear_regression/linear_regression.hpp:
```diff
@@ -87,11 +88,12 @@ class LinearRegression
    * @param responses y, the responses to the data points.
    * @param intercept Whether or not to fit an intercept term.
    * @param weights Observation weights (for boosting).
+   * @return ordinary least squares error.
```
Tiny style issue; we should start the comment capitalized, i.e. `Ordinary` :)

rcurtin commented:
Hehe, always bugs in code... I wonder why mlpack-bot offered stickers twice...

walragatver commented:
Hi @rcurtin,
I have updated History.md. However was not able to preview the change. So if something goes wrong you handle that during merge.
Thanks for approval :+1:.

rcurtin commented:
Thanks for the contribution! :+1:
